diff --git a/freepager.module b/freepager.module
index 2b7d8e2..11d4dee 100644
--- a/freepager.module
+++ b/freepager.module
@@ -63,25 +63,18 @@ function freepager_preprocess_views_freepager(&$vars) {
     $rendered_path = $view->render_field($field_names['path'], $row_number);
     if ($rendered_path == $current_path) {
 
-      // Prepare *all* the fields in this row – there may be token replacement
-      // patterns required by the fields we use. And also prepare the fields
-      // in the previous and next rows.
-      $rows_to_render = range($row_number - 1, $row_number + 1);
-      foreach ($rows_to_render as $row_to_render) {
-        foreach(array_keys($view->display_handler->get_field_labels()) as $field_name) {
-          $view->render_field($field_name, $row_to_render);
-        }
-      }
-
       if ($field_names['previous']) {
+      	_freepager_render_req_fields($view, $row_number - 1);
         $vars['previous'] = $view->render_field($field_names['previous'], $row_number - 1);
         $vars['previous_linked'] = l($vars['previous'], $view->render_field($field_names['path'], $row_number - 1), array('html' => TRUE));
       }
       if ($field_names['current']) {
+      	_freepager_render_req_fields($view, $row_number);
         $vars['current'] = $view->render_field($field_names['current'], $row_number);
         $vars['current_linked'] = l($vars['current'], $view->render_field($field_names['path'], $row_number), array('html' => TRUE));
       }
       if ($field_names['next']) {
+      	_freepager_render_req_fields($view, $row_number + 1);
         $vars['next'] = $view->render_field($field_names['next'], $row_number + 1);
         $vars['next_linked'] = l($vars['next'], $view->render_field($field_names['path'], $row_number + 1), array('html' => TRUE));
       }
@@ -100,6 +93,21 @@ function freepager_preprocess_views_freepager(&$vars) {
 }
 
 /**
+ * Internal function
+ * 
+ * Prepare *all* the fields in this row – there may be token replacement 
+ * patterns required by the fields we use. And also prepare the fields
+ * in the previous and next rows.
+ */
+function _freepager_render_req_fields($view, $row_to_render){
+	
+	foreach(array_keys($view->display_handler->get_field_labels()) as $field_name) {
+		$view->render_field($field_name, $row_to_render);
+	}
+	
+}
+
+/**
  * Implements hook_views_post_render().
  */
 function freepager_views_post_render(&$view, &$output, &$cache) {