diff --git a/includes/content.inc b/includes/content.inc
index 134c951..dcc5389 100644
--- a/includes/content.inc
+++ b/includes/content.inc
@@ -543,6 +543,25 @@ function ctools_content_configure_form_defaults_submit(&$form, &$form_state) {
  * such as whether or not you're using ajax, or the modal. $form_info will need
  * your next/submit callbacks so that you can cache your data appropriately.
  *
+ * @param $op
+ *   In which step of the content form are we. Possible values:
+ *    - 'add'
+ *    - others
+ * @param $form_info
+ *   The multistep form info of the ctools content form.
+ * @param $form_state
+ *   The current form_state on the form we are configuring at the moment.
+ * @param $plugin
+ *   The content type plugin info of the current content type.
+ * @param $subtype_name
+ *   The machine readable name of the current subtype of $plugin.
+ * @param $subtype
+ *   The subtype information of the current subtype of $plugin.
+ * @param $conf
+ *   The configuration options of the content type.
+ * @param $step
+ *   On which step of the multistep function are we.
+ *
  * @return
  *   If this function returns false, no form exists.
  */
diff --git a/includes/fields.inc b/includes/fields.inc
index dcd6139..0083d59 100644
--- a/includes/fields.inc
+++ b/includes/fields.inc
@@ -67,8 +67,10 @@ function ctools_fields_fake_field_instance($field_name, $view_mode = 'ctools', $
  *   The full form_state from the function that's calling this function.
  * @param $view_mode
  *   We're passing a view mode from this function to the fake instance we're creating.  Defaults to ctools, but we expect developers to actually name this something meaningful.
+ * @param $form_key
+ *   Sets the formatter settings to a certain key in the $form array.
  */
-function ctools_fields_get_field_formatter_settings_form($field, $formatter_type, &$form, $form_state, $view_mode = 'ctools') {
+function ctools_fields_get_field_formatter_settings_form($field, $formatter_type, &$form, $form_state, $view_mode = 'ctools', $form_key = '') {
   $conf = $form_state['conf'];
   $formatter = field_info_formatter_types($formatter_type);
   if (isset($formatter['settings'])) {
@@ -80,7 +82,12 @@ function ctools_fields_get_field_formatter_settings_form($field, $formatter_type
     $settings_form = $function($field, $instance, $view_mode, $form, $form_state);
     if ($settings_form) {
       $form['ctools_field_list']['#value'][] = $field;
-      $form += $settings_form;
+      if (!empty($form_key)) {
+        $form[$form_key] += $settings_form;
+      }
+      else {
+        $form += $settings_form;
+      }
     }
   }
 
diff --git a/plugins/content_types/entity_context/entity_field.inc b/plugins/content_types/entity_context/entity_field.inc
index acc5684..6d3650a 100644
--- a/plugins/content_types/entity_context/entity_field.inc
+++ b/plugins/content_types/entity_context/entity_field.inc
@@ -58,7 +58,6 @@ function ctools_entity_field_content_type_content_types() {
                   '@field_name' => $field_name,
                 )),
               ),
-              'ctools_entity_field_content_type_formatter_styles' => t('Formatter Styles'),
             ),
           );
         }
@@ -189,6 +188,7 @@ function ctools_entity_field_content_type_formatter_options($form, &$form_state)
     'hidden' => t('Hidden'),
   );
 
+
   $form['label'] = array(
     '#type' => 'select',
     '#title' => t('Label'),
@@ -201,37 +201,45 @@ function ctools_entity_field_content_type_formatter_options($form, &$form_state)
     '#title' => t('Select a formatter'),
     '#options' => $formatter_options,
     '#default_value' => $conf['formatter'],
+    '#ajax' => array(
+      'wrapper' => 'ctools-content-entity-field-formatter-settings',
+      'callback' => 'ctools_entity_field_content_type_formatter_change',
+    ),
+    '#submit' => array('ctools_entity_field_content_type_formatter_options_submit_temporary'),
+    '#executes_submit_callback' => TRUE,
   );
 
-  return $form;
-}
-
-function ctools_entity_field_content_type_formatter_options_submit($form, &$form_state) {
-  $form_state['conf']['formatter'] = $form_state['values']['formatter'];
-  $form_state['conf']['label'] = $form_state['values']['label'];
-}
-
-function ctools_entity_field_content_type_formatter_styles($form, &$form_state) {
-  if (!$form_state['conf']['formatter_settings']) {
-    $form_state['conf']['formatter_settings'] = array();
-  }
-  $conf = $form_state['conf'];
-  $subtype = $form_state['subtype_name'];
-  list($entity_type, $field_name) = explode(':', $subtype, 2);
-  $field = field_info_field($field_name);
-  module_load_include('inc', 'field_ui', 'field_ui.admin');
+  // Load the files to be able to execute the submit/ajax callback
+  $form_state['build_info']['files'][] = array(
+    'type' => 'inc',
+    'module' => 'ctools',
+    'name' => 'plugins/content_types/entity_context/entity_field',
+  );
+  $form_state['build_info']['files'][] = array(
+    'type' => 'inc',
+    'module' => 'ctools',
+    'name' => 'includes/fields',
+  );
 
-  ctools_include('fields');
   $form['ctools_field_list'] = array(
     '#type' => 'value',
     '#value' => array(),
   );
 
-  ctools_fields_get_field_formatter_settings_form($field, $conf['formatter'], $form, $form_state);
+  // Wrap the formatter settings in a new div, to allow an ajax update.
+  ctools_include('fields');
+  $form['settings']['#prefix'] = '<div id="ctools-content-entity-field-formatter-settings">';
+  ctools_fields_get_field_formatter_settings_form($field, $conf['formatter'], $form, $form_state, 'ctools', 'settings');
+  $form['settings']['#suffix'] = '</div>';
+
+
   return $form;
 }
 
-function ctools_entity_field_content_type_formatter_styles_submit($form, &$form_state) {
+function ctools_entity_field_content_type_formatter_options_submit($form, &$form_state) {
+  $form_state['conf']['formatter'] = $form_state['values']['formatter'];
+  $form_state['conf']['label'] = $form_state['values']['label'];
+
   $fields = $form_state['values']['ctools_field_list'];
   $formatter_info = ctools_fields_get_field_formatter_info($fields);
   foreach ($formatter_info as $info) {
@@ -259,3 +267,17 @@ function ctools_entity_field_content_type_admin_title($subtype, $conf, $context)
   ctools_include('fields');
   return t('"@s" @field', array('@s' => $context->identifier, '@field' => ctools_field_label($field_name)));
 }
+
+
+
+function ctools_entity_field_content_type_formatter_change($form, &$form_state) {
+  return $form['settings'];
+}
+
+function ctools_entity_field_content_type_formatter_options_submit_temporary($form, &$form_state) {
+  $form_state['conf']['formatter'] = $form_state['values']['formatter'];
+  $form_state['conf']['label'] = $form_state['values']['label'];
+
+  $form_state['rebuild'] = TRUE;
+  $form_state['rerender'] = TRUE;
+}
